### PR TITLE
Properly deinit of OpenSSL

### PR DESCRIPTION
**Pull request by jesferman**
I leave you two comments in code to discuss with you

---
This change is [Reviewable](https://reviewable.io/reviews/alanxz/rabbitmq-c/422)

### CODE REVIEW

File: librabbitmq/amqp_openssl.c [hunk 1]
```diff
@@ -662,8 +663,22 @@ destroy_openssl(void)
      * safely. We do leak the amqp_openssl_lockarray. Which is only
```
Comment (alanxz) [inline librabbitmq/amqp_openssl.c]: I'd like to remember why we're intensionally leaking `amqp_openssl_lockarray`.

This code was originally patterned after what the postgres client-library does, which had a reason for doing what it does, but I'm not currently remembering what it was.

File: librabbitmq/amqp_openssl.c [hunk 2]
```diff
@@ -662,8 +663,22 @@ destroy_openssl(void)
...
+    CONF_modules_free();
+    EVP_cleanup();
+    CRYPTO_cleanup_all_ex_data();
+    ERR_free_strings();
+    free(amqp_openssl_lockarray);
+    amqp_openssl_lockarray = NULL;
+    /* Allow the possibility to reinitialze openSSL again */
+    openssl_initialized = 0;
+    /* In newer versions of OpenSSL, call:
+     SSL_COMP_free_compression_methods();
```
Comment (alanxz) [inline librabbitmq/amqp_openssl.c]: Probably need to hack this up with `#ifdef`'s depending on what version of OpenSSL is being compiled against.

See: ~do_init() in http://www.boost.org/doc/libs/1_64_0/boost/asio/ssl/detail/impl/openssl_init.ipp

**Comment by jesferman** [comment]
Pull request updated after your review

**Comment by jesferman** [comment]
Hi @alanxz, I updated the code 18 days ago with the changes you proposed. Could you review it?

File: librabbitmq/amqp_openssl.c (hunk 1 above)
Comment (alanxz) [inline librabbitmq/amqp_openssl.c]: Still unclear as to why it was done this way; this change doesn't make the problem worse, so I'm inclined to leave this as is for now.

**Pull request merged**